### PR TITLE
extract some grandpa types to Primitives crate

### PR DESCRIPTION
**Pull request by yjhmelody**
- I improve type params since many types only want to use `Header` rather than the `Block` type.
- These types are used by me in both no_std/std. And  I need grandpa justification field be `pub`.

### CODE REVIEW

File: client/finality-grandpa/src/justification.rs [hunk 1]
```diff
@@ -42,9 +43,25 @@ use crate::{AuthorityList, Commit, Error};
 /// nodes, and are used by syncing nodes to prove authority set handoffs.
 #[derive(Clone, Encode, Decode, PartialEq, Eq, Debug)]
 pub struct GrandpaJustification<Block: BlockT> {
```
Comment (bkchr) [inline client/finality-grandpa/src/justification.rs:45]:
```suggestion
pub struct GrandpaJustification<Header: HeaderT> {
```

**Review by bkchr — APPROVED** [review]
Two nitpicks, otherwise it looks good.

File: client/finality-grandpa/src/justification.rs (hunk 1 above)
Comment (yjhmelody) [reply client/finality-grandpa/src/justification.rs:45]: I also want to use it in this way. But its impls must use Block type.

**Comment by davxy** [comment]
LGTM.

If we want to go even deeper I think that almost everything concerning communication
can be bounded over `Header` trait instead of `Block`.

For example starting from the `CurrentRounds`

https://github.com/yjhmelody/substrate/blob/6e4f14775ab528d5650ad0e6ed8220bb99601331/client/finality-grandpa/src/environment.rs#L173

The various messages

https://github.com/yjhmelody/substrate/blob/6e4f14775ab528d5650ad0e6ed8220bb99601331/client/finality-grandpa/src/communication/gossip.rs#L347

All the way up to the `GossipMessage`.

https://github.com/yjhmelody/substrate/blob/6e4f14775ab528d5650ad0e6ed8220bb99601331/client/finality-grandpa/src/communication/gossip.rs#L326

So in short I think that Grandpa mostly requires to be generic over the `Header` (as the external crate it is using).

**Should also be said that in practice it also doesn't make much difference.**
**Indeed in a lot of places (in Substrate) generics are bound to `Block`  where they can just be bound to `Header`.**

**Review by davxy — APPROVED** [review]

**Comment by yjhmelody** [comment]
> LGTM.
> 
> If we want to go even deeper I think that almost everything concerning communication can be bounded over `Header` trait instead of `Block`.
> 
> For example starting from the `CurrentRounds`
> 
> https://github.com/yjhmelody/substrate/blob/6e4f14775ab528d5650ad0e6ed8220bb99601331/client/finality-grandpa/src/environment.rs#L173
> 
> The various messages
> 
> https://github.com/yjhmelody/substrate/blob/6e4f14775ab528d5650ad0e6ed8220bb99601331/client/finality-grandpa/src/communication/gossip.rs#L347
> 
> All the way up to the `GossipMessage`.
> 
> https://github.com/yjhmelody/substrate/blob/6e4f14775ab528d5650ad0e6ed8220bb99601331/client/finality-grandpa/src/communication/gossip.rs#L326
> 
> So in short I think that Grandpa mostly requires to be generic over the `Header` (as the external crate it is using).
> 
> **Should also be said that in practice it also doesn't make much difference.** **Indeed in a lot of places (in Substrate) generics are bound to `Block` where they can just be bound to `Header`.**

Yeah, I know. Current I need to verifiy grandpa onchain without Block trait.

**Comment by davxy** [comment]
> > So in short I think that Grandpa mostly requires to be generic over the `Header` (as the external crate it is using).
> > **Should also be said that in practice it also doesn't make much difference.** **Indeed in a lot of places (in Substrate) generics are bound to `Block` where they can just be bound to `Header`.**
> 
> Yeah, I know. When grandpa types and logic are used for a specific instance chain, Block trait constraint is enough. But current I need to verifiy grandpa in multi-chains(though all Block types maybe are same). So Header trait is better.

Indeed I was silently pushing for bouding everything to `Header`  :-D
But that is a lot of work and maybe a subject for another PR

**Comment by bkchr** [comment]
bot merge
